# Models > Linear Regression

<div class="alert alert-info">The workhorse of empirical research in the social sciences</div>

In [2]:
# install latest version of pyrsm
# make sure to use --user
# %pip install --user pyrsm --upgrade

In [3]:
import matplotlib as mpl
import numpy as np

import pyrsm as rsm

# increase plot resolution
mpl.rcParams["figure.dpi"] = 100

In [4]:
rsm.__version__

'1.6.2'

In [5]:
## setup pyrsm for autoreload
%reload_ext autoreload
%autoreload 2
%aimport pyrsm

### Example 1: Catalog sales

We have access to data from a company selling men's and women's apparel through mail-order catalogs (dataset `catalog`). The company maintains a database on past and current customers' value and characteristics. Value is determined as the total \$ sales to the customer in the last year.

The catalog company is interested in redesigning their Customer Relationship Management (CRM) strategy. We will proceed in two steps:

1. Estimate a regression model using last year's sales total. Response variable: sales total for each of the 200 households; Explanatory variables: household income (measured in thousands of dollars), size of household, and age of the household head.
2. Interpret each of the estimated coefficients. Also provide a statistical evaluation of the model as a whole.
3. Which explanatory variables are significant predictors of customer value (use a 95% confidence level)?


In [6]:
# load example data from the pyrsm library
catalog, catalog_description = rsm.load_data(pkg="model", name="catalog")
catalog

id   Sales  Income  HH_size   Age
0      1  178.63    93.0        4  55.0
1      2  338.59    79.0        5  35.0
2      3  210.26    70.0        4  64.0
3      4  378.64    95.0        2  39.0
4      5  227.09   119.0        2  43.0
..   ...     ...     ...      ...   ...
195  196  234.85    42.0        2  38.0
196  197  138.20    94.0        2  58.0
197  198  340.74    88.0        2  35.0
198  199  496.48   110.0        5  37.0
199  200  111.84    31.0        1  44.0

[200 rows x 5 columns]

In [7]:
# format the data description
rsm.md(catalog_description)

## Catalog sales

### Description

Data from a company selling men's and women's apparel through mail-order catalogs. The company maintains a database on past and current customers' value and characteristics. Value is determine as the total $ sales to the customer in the last year. The data are a random sample of 200 customers from the company's database.

### Variables

A data frame with 200 observations on 4 variables

- id = Customer id
- Sales =	Total sales (in $) to a household in the past year
- Income = Household income ($1000) 
- HH.size = Size of the household (# of people) 
- Age = Age of the head of the household

In [8]:
reg = rsm.model.regress({"catalog": catalog}, rvar="Sales", evar=["Income", "HH_size", "Age"])

In [9]:
reg.summary(rmse=True, ssq=True)

Linear regression (OLS)
Data                 : catalog
Response variable    : Sales
Explanatory variables: Income, HH_size, Age
Null hyp.: the effect of x on Sales is zero
Alt. hyp.: the effect of x on Sales is not zero

           coefficient  std.error  t.value p.value     
Intercept       45.359     43.089    1.053   0.294     
Income           1.775      0.217    8.181  < .001  ***
HH_size         22.122      4.190    5.279  < .001  ***
Age              0.449      0.768    0.585   0.559     

Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

R-squared: 0.331, Adjusted R-squared: 0.321
F-statistic: 32.325 df(3, 196), p.value < 0.001
Nr obs: 200

Sum of squares:

             df         SS
Regression    3    770,565
Error       196  1,557,415
Total       199  2,327,980

Root Mean Square Error (RMSE):
88.244


The null and alternate hypothesis for the F-test can be formulated as follows:

* $H_0$: All regression coefficients are equal to 0
* $H_a$: At least one regression coefficient is not equal to zero

The F-statistic suggests that the regression model as a whole explains a significant amount of variance in `Sales`. The calculated F-statistic is equal to 32.33 and has a very small p.value (< 0.001). The amount of variance in sales explained by the model is equal to 33.1% (see R-squared).

We can replicate the standard F-test that is reported as part of all regression output by passing `income`, `HH.size`, and `Age` to the `f_test` method. The relevant output is shown below.

In [10]:
reg.f_test(["Income", "HH_size", "Age"])


Model 1: Sales ~ 1
Model 2: Sales ~ Income + HH_size + Age
R-squared, Model 1 vs 2: 0.000 vs 0.331
F-statistic: 32.325 df (3, 196), p.value < .001


Note that in this example, "model 1" is a regression without explanatory variables. As you might expect, the explained variance for model 1 is equal to zero. The F-test compares the _fit_ of model 1 and model 2, adjusted for the difference in the number of coefficients estimated in each model. The test statistic to use is described below. $R^2_2$ is the explained variance for model 2 and $R^2_1$ is the explained variance for model 1. $n$ is equal to the number of rows in the data, and $k_2$ ($k_1$) is equal to the number of estimated coefficients in model 2 (model 1).

$$
\begin{align}
  F & = \frac{(R^2_2 - R^2_1)/(k_2 - k_1)}{(1 - R^2_2)/(n - k_2 - 1)} \\
    & = \frac{(0.331 - 0)/(3 - 0)}{(1 - 0.331)/(200 - 3 - 1)} \\
    & = 32.325
\end{align}
$$

We can use the provided p.value associated with an F-statistic of 32.325 to evaluate the null hypothesis. We can also calculate the critical F-statistic using the probability calculator. As we can see from the output below that value is 2.651. Because the provided p.value is < 0.001 and the calculated F-statistic is larger than the critical value (32.325 > 2.651) we reject null the hypothesis that all coefficient are equal to zero.

In [11]:
pc = rsm.basics.prob_calc("fdist", df1=3, df2=196, pub=0.95)
pc.summary()
pc.plot()

Probability Calculator
Distribution: F
Df 1        : 3
Df 2        : 196
Mean        : 1.01
Variance    : 0.698
Lower bound : 0
Upper bound : 0.95

P(X < 2.651) = 0.95
P(X > 2.651) = 0.05


The coefficients from the regression can be interpreted as follows:

- For an increase in income of \$1,000 we expect, on average, to see an increase in sales of 1.7754 dollars, keeping all other variables in the model constant.
- For an increase in household size of 1 person we expect, on average, to see an increase in sales of \$22.1218, keeping all other variables in the model constant.
- For an increase in the age of the head of the household of 1 year we expect, on average, to see an increase in sales of \$0.45, keeping all other variables in the model constant.

For each of the explanatory variables the following null and alternate hypotheses can be formulated:

* H0: The coefficient associated with explanatory variable x is equal to 0
* Ha: The coefficient associated with explanatory variable x is not equal to 0

The coefficients for `Income` and `HH.size` are both significant (p.values < 0.05), i.e., we can reject H0 for each of these coefficients. The coefficient for `Age HH` is not significant (p.value > 0.05), i.e., we cannot reject H0 for `Age HH`. We conclude that a change in Age of the household head does not lead to a significant change in sales.

We can also use the t.values to evaluate the null and alternative hypotheses for the coefficients. Because the calculated t.value for `Income` and `HH.size` is **larger** than the _critical_ t.value we reject the null hypothesis for both effects. We can obtain the critical t.value by using the probability calculator in the _Basics_ menu. For a t-distribution with 196 degrees of freedom (see the `Errors` line in the `Sum of Squares` table shown above) the critical t.value is 1.972. We have to enter 0.025 and 0.975 as the lower and upper probability bounds in the probability calculator because the alternative hypothesis is `two.sided`.

In [12]:
pc = rsm.basics.prob_calc("tdist", df=196, plb=0.025, pub=0.975)
pc.summary()
pc.plot()

Probability Calculator
Distribution             : t
Df                       : 196
Mean                     : 0
St. dev                  : 1.01
Lower bound              : 0.025
Upper bound              : 0.975
1 - P(-1.972 < X < 1.972): 0.05

P(X < -1.972) =       0.025
P(X > -1.972) =       0.975
P(X < 1.972) =        0.975
P(X > 1.972) =        0.025
P(-1.972 < X < 1.972) = 0.95


### Example 2: Ideal data for regression

The data `ideal` contains simulated data that is very useful to demonstrate what data for, and residuals from, a regression should ideally look like. The data has 1,000 observations on 4 variables. `y` is the response variable and `x1`, `x2`, and `x3` are explanatory variables. The plots shown below can be used as a bench mark for regressions on real world data

In [13]:
ideal, ideal_description = rsm.load_data(pkg="model", name="ideal")
ideal

y        x1         x2        x3
0    -5.975681  0.152841   5.011979  3.155253
1     0.958703  0.190931   3.233194  5.354473
2    -2.031790  0.588247   4.891027  5.832465
3    -7.318411  0.268413   7.682724  6.600717
4    -2.746885  0.672403   5.358413  3.147829
..         ...       ...        ...       ...
995 -14.565060  0.122622  10.439340  7.018642
996  -2.374663  0.875633   5.491546  5.039407
997  -4.267488  0.517993   8.487617  6.668247
998  -0.511373  0.843896   6.420335  3.693862
999  -5.187129  0.105630   5.806936  4.022727

[1000 rows x 4 columns]

In [14]:
rsm.md(ideal_description)

## 'Ideal' data for regression

### Description

Simulated data set used to illustrate evaluation of regression residuals.  

### Variables

A data frame with 1000 observations on 4 variables (y, x1, x2, and x3)

`y`, `x2`, and `x3` appear (roughly) normally distributed whereas `x1` appears (roughly) uniformly distributed. There are no indication of outliers or severely skewed distributions.

In [15]:
reg = rsm.model.regress({"ideal": ideal}, rvar="y", evar=["x1", "x2", "x3"])

In [16]:
reg.plot("dist")

In the plot of correlations there are clear associations among the response and explanatory variables as well as among the explanatory variables themselves. Note that in an experiment the x’s of interest would have a zero correlation. This is very unlikely in historical data however. The scatter plots in the lower-diagonal part of the plot show that the relationships between the variables are (approximately) linear.

In [17]:
reg.plot(plots="corr")

The scatter plots of `y` (the response variable) against each of the explanatory variables confirm the insight from the correlation plot. 

In [18]:
reg.summary()

Linear regression (OLS)
Data                 : ideal
Response variable    : y
Explanatory variables: x1, x2, x3
Null hyp.: the effect of x on y is zero
Alt. hyp.: the effect of x on y is not zero

           coefficient  std.error  t.value p.value     
Intercept        0.669      0.279    2.398   0.017    *
x1               7.713      0.213   36.157  < .001  ***
x2              -1.987      0.043  -46.369  < .001  ***
x3               0.902      0.070   12.861  < .001  ***

Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

R-squared: 0.804, Adjusted R-squared: 0.803
F-statistic: 1360.646 df(3, 996), p.value < 0.001
Nr obs: 1,000


In [19]:
reg.plot("scatter")

The dashboard of six residual plots looks excellent, as we might expect for these data. True values and predicted values from the regression form a straight line with random scatter, i.e., as the actual values of the response variable go up, so do the predicted values from the model. The residuals (i.e., the differences between the values of the response variable data and the values predicted by the regression) show no pattern and are randomly scattered around a horizontal line. Any pattern would suggest that the model is better (or worse) at predicting some parts of the data compared to others. If a pattern were visible in the Residual vs Row order plot we might be concerned about auto-correlation. Again, the residuals are nicely scattered about a horizontal axis. Note that auto-correlation is a problem we are really only concerned about when we have time-series data. The Q-Q plot shows a nice straight and diagonal line, evidence that the residuals are normally distributed. This conclusion is confirmed by the histogram of the residuals and the density plot of the residuals (green) versus the theoretical density of a normally distributed variable (blue line).

In [20]:
reg.plot("dashboard")

The final diagnostic we will discuss is a set of plots of the residuals versus the explanatory variables (or predictors). There is no indication of any trends or heteroscedasticity. Any patterns in these plots would be cause for concern. There are also no outliers, i.e., points that are far from the main cloud of data points.

In [21]:
reg.plot("residual")

Since the diagnostics look good, we can draw inferences from the regression. First, the model is significant as a whole: the p.value on the F-statistic is less than 0.05 therefore we reject the null hypothesis that all three variables in the regression have slope equal to zero. Second, each variable is statistically significant. For example, the p.value on the t-statistic for x1 is less than 0.05 therefore we reject the null hypothesis that x1 has a slope equal to zero when x2 and x3 are also in the model (i.e., ‘holding all other variables in the model constant’).

Increases in x1 and x3 are associated with increases in y whereas increases in x2 are associated with decreases in y. Since these are simulated data the exact interpretation of the coefficient is not very interesting. However, in the scatterplot it looks like increases in x3 are associated with decreases in y. What explains the difference? Hint: consider the correlation plots.

In [22]:
reg.summary()

Linear regression (OLS)
Data                 : ideal
Response variable    : y
Explanatory variables: x1, x2, x3
Null hyp.: the effect of x on y is zero
Alt. hyp.: the effect of x on y is not zero

           coefficient  std.error  t.value p.value     
Intercept        0.669      0.279    2.398   0.017    *
x1               7.713      0.213   36.157  < .001  ***
x2              -1.987      0.043  -46.369  < .001  ***
x3               0.902      0.070   12.861  < .001  ***

Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

R-squared: 0.804, Adjusted R-squared: 0.803
F-statistic: 1360.646 df(3, 996), p.value < 0.001
Nr obs: 1,000


### Example 3: Linear or log-log regression?

Both linear and log-log regressions are commonly applied to business data. In this example we will look for evidence in the data and residuals that may suggest which model specification is  appropriate for the available data.

In [23]:
diamonds, diamonds_description = rsm.load_data(pkg="data", name="diamonds")
diamonds

price  carat clarity        cut color  depth  table     x     y     z  \
0       580   0.32     VS1      Ideal     H   61.0   56.0  4.43  4.45  2.71   
1       650   0.34     SI1  Very Good     G   63.4   57.0  4.45  4.42  2.81   
2       630   0.30     VS2  Very Good     G   63.1   58.0  4.27  4.23  2.68   
3       706   0.35    VVS2      Ideal     H   59.2   56.0  4.60  4.65  2.74   
4      1080   0.40     VS2    Premium     F   62.6   58.0  4.72  4.68  2.94   
...     ...    ...     ...        ...   ...    ...    ...   ...   ...   ...   
2995   4173   1.14     SI1  Very Good     J   63.3   55.0  6.60  6.67  4.20   
2996   8396   1.51     SI1      Ideal     I   61.2   60.0  7.39  7.37  4.52   
2997    449   0.32     VS2    Premium     I   62.6   58.0  4.37  4.42  2.75   
2998   4370   0.91     VS1  Very Good     H   62.1   59.0  6.17  6.20  3.84   
2999    453   0.26     VS2  Very Good     E   63.2   56.0  4.06  4.10  2.58   

            date  
0     2012-02-26  
1     2012-02-26  
2     2012-02-26  
3     2012-02-26  
4     2012-02-26  
...          ...  
2995  2015-12-01  
2996  2015-12-01  
2997  2015-12-01  
2998  2015-12-01  
2999  2015-12-01  

[3000 rows x 11 columns]

The data diamonds contains information on prices of 3,000 diamonds. A more complete description of the data and variables is available below.

In [24]:
rsm.md(diamonds_description)

## Diamond prices

Prices of 3,000 round cut diamonds

### Description

A dataset containing the prices and other attributes of a sample of 3000 diamonds. The variables are as follows:

### Variables

- price = price in US dollars ($338--$18,791)
- carat = weight of the diamond (0.2--3.00)
- clarity = a measurement of how clear the diamond is (I1 (worst), SI2, SI1, VS2, VS1, VVS2, VVS1, IF (best))
- cut = quality of the cut (Fair, Good, Very Good, Premium, Ideal)
- color = diamond color, from J (worst) to D (best)
- depth = total depth percentage = z / mean(x, y) = 2 * z / (x + y) (54.2--70.80)
- table = width of top of diamond relative to widest point (50--69)
- x = length in mm (3.73--9.42)
- y = width in mm (3.71--9.29)
- z = depth in mm (2.33--5.58)
- date = shipment date

### Additional information

<a href="http://www.diamondse.info/diamonds-clarity.asp" target="_blank">Diamond search engine</a>


 Select the variable price as the response variable and carat and clarity as the explanatory variables. Before looking at the parameter estimates from the regression go to the Plots tab to take a look at the data and residuals. Below are the histograms for the variables in the model. Price and carat seem skewed to the right. Note that the direction of skew is determined by where the tail is.

In [25]:
reg = rsm.model.regress({"diamonds": diamonds}, rvar="price", evar=["carat", "clarity"])

In [26]:
reg.plot("dist")

In [27]:
reg.plot("corr")

The scatter plots of price (the response variable) against the explanatory variables are not as clean as for the ideal data in Example 2. In fact the points seem to fan-out for higher prices and number of carats. 

In [28]:
reg.plot("scatter", nobs=-1)

There does not seem to be very much movement in price for different levels of clarity. If anything, the price of the diamond seems to go down as clarity increase. A surprising result we will discuss in more detail below.

The dashboard of six residual plots looks less than stellar. The actual values and fitted (predicted) values from the regression show curvature. At higher actual and fitted values the spread of points around the line is wider, consistent with what we saw in the scatter plot of price versus carat. The residuals (i.e., the differences between the actual data and the values predicted by the regression) show an even more distinct pattern as they are clearly not randomly scattered around a horizontal axis. The Residual vs Row order plot looks perfectly straight indicating that auto-correlation is not a concern. Finally, while for the ideal data in Example 2 the Q-Q plot showed a nice straight diagonal line, here dots clearly separate from the line at the right extreme. Evidence that the residuals are not normally distributed. This conclusions is confirmed by the histogram and density plots of the residuals that show a more spiked appearance than a normally distributed variable would.

In [29]:
reg.plot("dashboard")

The final diagnostic we will discuss is a set of plots of the residuals versus the explanatory variables (or predictors). The residuals fan-out from left to right in the plot of residuals vs carats. The scatter plot of `clarity` versus residuals shows outliers with strong negative values for lower levels of `clarity` and outliers with strong positive values for diamonds with higher levels of `clarity`.

In [30]:
reg.plot("residual")

Since the diagnostics do not look good, we should not draw conclusions /(inference) from this regression. A log-log specification may be preferable.

In [31]:
diamonds = diamonds.assign(price_ln=np.log(diamonds.price), carat_ln=np.log(diamonds.carat))

In [32]:
reg = rsm.model.regress({"diamonds": diamonds}, rvar="price_ln", evar=["carat_ln", "clarity"])

The relationship between log-price and log-carat looks close to linear. Exactly what we are looking for!

In [33]:
reg.plot("scatter", nobs=-1)

We will apply a (natural) log (or ln) transformation to both price and carat and rerun the analysis to see if the log-log specification is more appropriate for the available data. This transformation can be done in Data > Transform. Select the variables price and carat. Choose Transform from the Transformation type drop-down and choose Ln (natural log) from the Apply function drop-down. Make sure to click the Store button so the new variables are added to the dataset. Note that we cannot apply a log transformation to clarity because it is a categorical variable.

Specify `price_ln` as the response variable and `carat_ln` and `clarity` as the explanatory variables. Before looking at the parameter estimates from the regression lets take a look at the data and residuals. Below are the histograms for the variables in the model. Note that `price_ln` and `carat_ln` are no longer right skewed, a good sign.

In [34]:
reg.plot("dist")

In the plot of correlations there are still clear associations among the response and explanatory variables. The correlation between `price_ln` and `carat_ln` is extremely large (i.e., .93). The correlation between `carat_ln` and `clarity` of the diamond is significant and negative.

In [35]:
reg.plot("corr")

The scatter plots of price_ln (the response variable) against the explanatory variables are now much cleaner. The line through the scatter plot of price_ln versus carat_ln is (mostly) straight. Although the points do have a bit of a blocked shape around the line, the scattering seems mostly random. We no longer see the points fan-out for higher values of price_ln and carat_ln. There seems to be a bit more movement in price_ln for different levels of clarity. However, the price_ln of the diamond still goes down as clarity increases which is unexpected. We will discuss this result below.

In [36]:
reg.plot("scatter", nobs=-1)

The dashboard of six residual plots looks much better than for the linear model. The true values and predicted values from the regression (almost) form a straight line. Although at higher and lower actual and predicted values the line is perhaps still very slightly curved. The residuals are much closer to a random scatter around a horizontal line. The Residual vs Row order plot still looks perfectly straight indicating that auto-correlation is not a concern. Finally, the Q-Q plot shows a nice straight and diagonal line, just like we saw for the ideal data in Example 2. Evidence that the residuals are now normally distributed. This conclusion is confirmed by the histogram and density plot of the residuals.

In [37]:
reg.plot("dashboard")

The final diagnostic we will discuss is a set of plots of the residuals versus the explanatory variables (or predictors). The residuals look much closer to random scatter around a horizontal line compared to the linear model. Although for low (high) values of carat_ln the residuals may be a bit higher (lower).

In [38]:
reg.plot("vimp")

In [39]:
vimp_values = reg.plot("vimp", ret=True)
vimp_values

Importance
carat_ln    0.964192
clarity     0.064764

In [40]:
reg.plot("residual")

Since the diagnostics now look much better, we can feel more confident about drawing inferences from this regression. The regression results are available in the Summary tab. Note that we get 7 coefficients for the variable clarity compared to only one for carat_ln. How come? If you look at the data description (Data > Manage) you will see that clarity is a categorical variables with levels that go from IF (worst clarity) to I1 (best clarity). Categorical variables must be converted to a set of dummy (or indicator) variables before we can apply numerical analysis tools like regression. Each dummy indicates if a particular diamond has a particular clarity level (=1) or not (=0). Interestingly, to capture all information in the 8-level clarity variable we only need 7 dummy variables. Note there is no dummy variable for the clarity level I1 because we don’t actually need it in the regression. When a diamond is not of clarity SI2, SI1, VS2, VS1, VVS2, VVS1 or IF we know that in our data it must be of clarity I1.

The F-statistic suggests that the regression model as a whole explains a significant amount of variance in price_ln. The F-statistic is very large and has a very small p.value (< 0.001) so we can reject the null hypothesis that all regression coefficients are equal to zero. The amount of variance in price_ln explained by the model is equal to 96.6. It seems likely that prices of diamonds will be much easier to predict than demand for diamonds.

The null and alternate hypothesis for the F-test can be formulated as follows: H0: All regression coefficients are equal to 0 Ha: At least one regression coefficient is not equal to zero

The coefficients from the regression can be interpreted as follows:

* For a 1% increase in carats we expect, on average, to see a 1.809% increase in the price of a diamond of, keeping all other variables in the model constant
* Compared to a diamond of clarity I1 we expect, on average, to pay 100x(exp(.444)-1) = 55.89% more for a diamond of clarity SI2, keeping all other variables in the model constant
* Compared to a diamond of clarity I1 we expect, on average, to pay 100x(exp(.591)-1) = 80.58% more for a diamond of clarity SI1, keeping all other variables in the model constant
* Compared to a diamond of clarity I1 we expect, on average, to pay 100x(exp(1.080)-1) = 194.47% more for a diamond of clarity IF, keeping all other variables in the model constant

The coefficients for each of the levels of clarity imply that an increase in clarity will increase the price of diamond. Why then did the scatter plot of clarity versus (log) price show price decreasing with clarity? The difference is that in a regression we can determine the effect of a change in one variable (e.g., clarity) keeping all other variables in the model constant (i.e., carat). Bigger, heavier, diamonds are more likely to have flaws compared to small diamonds so when we look at the scatter plot we are really seeing the effect of not only improving clarity on price but also the effect of carats which are negatively correlated with clarity. In a regression, we can compare the effects of different levels of clarity on (log) price for a diamond of the same size (i.e., keeping carat constant). Without (log) carat in the model the estimated effect of clarity would be incorrect due to omitted variable bias. In fact, from a regression of price_ln on clarity we would conclude that a diamond of the highest clarity in the data (IF) would cost 59.22% less compared to a diamond of the lowest clarity (I1). Clearly this is not a sensible conclusion.

For each of the explanatory variables the following null and alternate hypotheses can be formulated: H0: The coefficient associated with explanatory variable X is equal to 0 Ha: The coefficient associated with explanatory variable X is not equal to 0

All coefficients in this regression are highly significant.

In [41]:
reg = rsm.model.regress({"diamonds": diamonds}, rvar="price_ln", evar=["carat_ln", "clarity"])
reg.summary()

Linear regression (OLS)
Data                 : diamonds
Response variable    : price_ln
Explanatory variables: carat_ln, clarity
Null hyp.: the effect of x on price_ln is zero
Alt. hyp.: the effect of x on price_ln is not zero

               coefficient  std.error  t.value p.value     
Intercept            7.802      0.030  263.458  < .001  ***
clarity[SI2]         0.444      0.031   14.463  < .001  ***
clarity[SI1]         0.591      0.031   19.332  < .001  ***
clarity[VS2]         0.749      0.031   24.348  < .001  ***
clarity[VS1]         0.792      0.031   25.376  < .001  ***
clarity[VVS2]        0.946      0.032   29.449  < .001  ***
clarity[VVS1]        1.011      0.033   30.866  < .001  ***
clarity[IF]          1.080      0.036   30.230  < .001  ***
carat_ln             1.809      0.006  284.096  < .001  ***

Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

R-squared: 0.966, Adjusted R-squared: 0.966
F-statistic: 10688.499 df(8, 2991), p.value < 0.001
Nr obs: 3,0

In [42]:
reg.predict(ci=True, conf=0.05).round(3)

carat_ln clarity  prediction  47.50%  52.50%
0       -1.139     VS1       6.533   6.532   6.534
1       -1.079     SI1       6.442   6.441   6.442
2       -1.204     VS2       6.373   6.373   6.374
3       -1.050    VVS2       6.849   6.848   6.850
4       -0.916     VS2       6.894   6.893   6.894
...        ...     ...         ...     ...     ...
2995     0.131     SI1       8.631   8.630   8.631
2996     0.412     SI1       9.139   9.139   9.140
2997    -1.139     VS2       6.490   6.489   6.490
2998    -0.094     VS1       8.424   8.423   8.424
2999    -1.347     VS2       6.114   6.114   6.115

[3000 rows x 5 columns]

In [43]:
reg.plot("coef")

/home/vnijs/gh/pyrsm/pyrsm/model/model.py:576: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`). Consider using `matplotlib.pyplot.close()`.
  fig = plt.figure(figsize=figsize)


# Radiant for Python App: Linear regression (OLS)

All the output shown above can be reproduced using the Radiant-for-Python web interface. An example of what the code required to start the web interface is shown below. See if you can reproduce the result.

> Note: The app will continue running until you press the `Stop` button in the app navigation bar or the stop button next to the notebook cell

In [ ]:
rsm.radiant.model.regress({"diamonds": diamonds}, {"diamonds": diamonds_description})

Listening on http://0.0.0.0:8000
Pyrsm and Radiant are open source tools and free to use. If you
are a student or instructor using pyrsm or Radiant for a class,
as a favor to the developers, please send an email to
<radiant@rady.ucsd.edu> with the name of the school and class.
If you are using Radiant in your company, as a favor to the
developer, please share the name of your company and what types
of activites you are supporting with the tool.
Opening in existing browser session.


```python
import pyrsm as rsm

# diamonds = pd.read_parquet('diamonds.parquet')
reg = rsm.model.regress(
    data={"diamonds": diamonds}, rvar="price", evar=["carat", "clarity", "cut"]
)
reg.summary()

```

<p align="center">
<img src="figures/linear-regression-summary.png">
</p>

<p align="center">
<img src="figures/linear-regression-pred-plots.png">
</p>